File: 프로그래머스/2/131704. 택배상자/택배상자.cpp
#include <string>
#include <vector>
#include <stack>
using namespace std;

int solution(vector<int> order) {
    int answer = 0;
    stack<int> stk;

    for (int i = 1; i <= order.size(); i++) {
        stk.push(i);

        while (!stk.empty() && stk.top() == order[answer]) {
            stk.pop();
            answer++;
        }
    }

    return answer;
}

File: 프로그래머스/2/131704. 택배상자/택배상자.java
import java.util.*;

class Solution {

    public int solution(int[] arr) {
        int answer = 0;
        Queue<Integer> main = new LinkedList<>();
        Deque<Integer> sub = new ArrayDeque<>();
        
        for (int i = 1; i <= arr.length; i++) {
            main.offer(i);
        }

        int i = 0;
        while (i < arr.length) {
            int should = arr[i];
            
            
            if (!main.isEmpty() && main.peek() != should) {
                if (!sub.isEmpty() && sub.peek() != should) {
                    sub.push(main.poll());
                } else if (!sub.isEmpty() && sub.peek() == should) {
                    sub.pop();
                    answer++;
                    i++;
                } else {
                    sub.push(main.poll());
                }
            } else if (!main.isEmpty() && main.peek() == should) {
                main.poll();
                answer++;
                i++;
            } else if (!sub.isEmpty() && sub.peek() == should) {
                sub.pop();
                answer++;
                i++;
            } else {
                break;
            }
        }
        
        return answer;
    }
}